torch/distributed/_tensor/placement_types.py

```diff
@@ -390,7 +390,13 @@ def __eq__(self, __o: object) -> bool:
             isinstance(__o, DTensorSpec)
             and self.mesh == __o.mesh
             and self.placements == __o.placements
-            and self.tensor_meta == __o.tensor_meta
+            # perf hack to avoid redistribute due to memory_format to be None.
+            and self.tensor_meta.shape == __o.tensor_meta.shape
```

Comment:
probably we should refactor tensor_meta to only take tensor shape/stride, and drop other fields (sharding prop only need shape anyways). This can be done in follow up PRs.

Reply:
I see. A good point. Do we want to keep the rest in a separate field or we want to completely drop it here? because this metadata is defined in torch.fx. Maybe we only compare the shape here?

Reply:
In order to do this, I think we can introduce another data structure (maybe call it ShapeSpec and only save shape/stride), then not use the metadata from torch.fx.

If we want to not get rid of the metadata in this PR, we can still only compare the shape here in dtensor spec `__eq__`